### PR TITLE
Add optional --file-durations flag for duration-aware bin-packing

### PR DESCRIPTION
**Pull request by eyalr**
## What

Adds an **opt-in** `--file-durations=<path>` flag. `<path>` is a JSON file
shaped like `{relative_file_path: total_seconds}`, using the exact same
relative-path convention `hash_filename()` has always used (strip the cwd,
or failing that a site-packages prefix, off the realpath).

When a valid file is given, nose-picker bin-packs that file set across
`--total-processes` bins by descending duration (greedy LPT -- longest
processing time first), so shards end up balanced by *total runtime*
instead of by *file count*. `wantFile()` becomes a simple membership check
against the bin assigned to `self.which_process`, computed once in
`configure()` (before collection starts), instead of a per-file hash
computation.

**Nose's own real discovery is untouched.** nose-picker does **not** walk
the filesystem itself to build a candidate list -- the candidate set for
bin-packing is simply the durations table's own key set (real ground truth
from a prior actual run's junit output), and nose's own `Loader`/`Selector`
keeps calling `wantFile()` once per file, one at a time, exactly as it
always has (see "Design evolution" below for why an earlier version of
this PR did try to walk the filesystem itself, and why that was dropped).
A file nose visits that the durations table doesn't cover (new since the
table was last refreshed, or for any other reason) falls back to the
classic hash for that one file, so it still runs in exactly one shard
rather than vanishing from all of them.

## Backward compatibility (the important part)

If `--file-durations` is **not passed at all**, or the given path doesn't
exist, isn't readable, or doesn't parse as valid JSON, behavior is
**exactly** the prior `hash_filename(name) % total_processes` selection,
byte-for-byte unchanged. This is deliberate and load-bearing: nose-picker
has other consumers besides the one motivating this change, and none of
them should see any behavior difference unless they explicitly opt in.

The entire duration-mode configuration path is wrapped so that *any*
unexpected failure degrades to the classic hash-based selection rather than
propagating and breaking CI for everyone. There's also a per-file safety
net (described above): an individual file the durations table doesn't cover
falls back to the hash for just that file.

Verified with an integration test that simulates all N shards at once and
checks the file set gets an exact 1:1 (no missing, no duplicated)
assignment across shards -- in legacy hash mode, in full duration mode,
with a partially/wholly-unmatched durations table, and in every fallback
path (missing file, invalid JSON, all-zero/NaN durations) -- and that those
fallback paths produce an assignment *identical* to pure legacy mode.

## Known ceiling -- please read before over-trusting this

Bin-packing here only ever operates at **whole-file granularity**, because
nose's `wantFile` hook never sees inside a file -- there's no mechanism to
split one file's tests across shards. If a single test file is much slower
than an even share of the whole suite, no bin-packing scheme built on this
hook can even out wall-clock time across shards; that one file's duration
becomes a floor for whichever shard it lands in. This helps a lot when
slowness is spread across many files (the common case), much less when
it's concentrated in one pathologically slow file (the fix there is
splitting that file, not a smarter scheduler).

## Design evolution during review (kept here for the audit trail)

This PR went through a meaningful redesign based on review feedback, plus
three bugbot-flagged-and-fixed rounds. Summarized here since the commit
history has all the detail:

1. **v1**: nose-picker did its own filesystem walk in `configure()`,
   reimplementing nose's *default* discovery convention (`testMatch`/
   `ignoreFiles`/`srcDirs`/package detection) to build a complete candidate
   list up front, since `wantFile()` alone never exposes the full set.
2. **Bugbot round 1**: a durations table with no usable timing signal at
   all (empty overlap, or every covered value `<= 0`) collapsed the entire
   LPT bin-packing loop onto shard 0 -- fixed by having the bin-packer
   signal "can't bin-pack" and fall back to hash mode, plus a follow-up fix
   for a narrower version of the same clumping bug with partial (not just
   total) zero-weight files, via a fairer tie-break rule.
3. **Bugbot round 2**: found a real symlink-duplication bug (fixed --
   `discover_candidate_files()` now deduped by `os.path.realpath()`) and a
   claimed basename-vs-full-path discovery mismatch that, on empirical
   verification against real nose 1.3.7 (not just source-reading), turned
   out to be a false positive -- nose really does only match a bare
   basename, confirmed by three separate fixture runs through actual
   `nosetests -v`. Documented, not "fixed" (the suggested change would have
   made discovery *more* permissive than real nose).
4. **Architectural change**: rather than keep hardening a from-scratch
   reimplementation of nose's discovery rules (however well-verified),
   dropped it entirely. The risk that motivated this -- any future
   divergence between the reimplementation and nose's actual behavior could
   silently drop a file from every shard, a correctness bug, not just a
   balance one -- goes away completely if nose-picker never tries to
   rediscover files itself. The candidate set is now just the durations
   table's own keys; nose's real discovery drives everything else, same as
   the original 0.5.7. Test suite rewritten to match (no more filesystem
   fixtures needed to exercise bin-packing at all).
5. **Bugbot round 3**: found that Python's `json` module accepts bare
   `NaN`/`Infinity`/`-Infinity` literals by default, and a `NaN` duration
   value broke the same downstream comparisons as round 1's bug (NaN is
   never `<`, `>`, or `==` anything, including itself) -- fixed by rejecting
   non-finite values at ingestion, plus defensively inside the bin-packer
   itself.

Current state: 28 tests, all passing under real nose 1.3.7 on Python 2.7.

---

> [!NOTE]
> **Medium Risk**
> Opt-in sharding logic is correctness-sensitive (every test file must map to exactly one shard), but failures fall back to legacy hashing and the PR adds broad regression coverage for collapse/imbalance cases.
> 
> **Overview**
> **0.6.0** adds an opt-in **`--file-durations`** JSON path so parallel shards can be balanced by **total runtime** (greedy LPT bin-packing in `configure()`) instead of only hash-modulo file count. Bin-packing uses the durations table’s keys as the candidate set—nose still drives discovery via `wantFile()`; known files are membership-checked per shard, and files not in the table fall back to the classic hash so each file still lands on exactly one shard.
> 
> Path handling is refactored through **`_relative_key()`** so JSON keys match **`hash_filename()`**. Invalid or missing duration data, unusable timing (e.g. all zeros/NaN), and configure errors all **degrade to unchanged hash behavior** when the flag is omitted or broken. **`report()`** can warn when too many test-like visited files were unknown to the table; a reporting-only **`_looks_testlike`** heuristic avoids noise from non-test `.py` files.
> 
> README documents the feature and limits (whole-file granularity). A large new **`tests/test_nose_plugin.py`** covers bin-packing edge cases, fallbacks, shard bijection, and staleness warnings.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit edeb49dbb140a28df0c6ddc7dc2c4a5ff9f0a431. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by eyalr** [comment]
bugbot run

File: src/picker/nose_plugin.py [hunk 1]
```diff
@@ -31,46 +31,292 @@
...
+        return False
+    if not basename.endswith('.py'):
+        return False
+    return bool(_DEFAULT_TESTMATCH_RE.search(basename))
+
+
+def _wants_directory(basename, fullpath):
+    if _is_package_dir(fullpath):
+        return True
+    return bool(_DEFAULT_TESTMATCH_RE.search(basename)) or basename in _DEFAULT_SRC_DIRS
```
[the remark this answers is not in the document]
Comment (eyalr) [reply src/picker/nose_plugin.py]: Investigated this against real nose 1.3.7 (not just by re-reading the source) before deciding whether to apply the suggested fix, since getting this wrong either way has real consequences.

Reproduced with two fixture layouts, run through actual `nosetests -v`:

1. `root/stuff/test_nested.py` (`stuff`: non-package, no "test" in its name, no matching ancestor either) -> `nosetests -v` collects 0 tests.
2. `root/tests/unit/test_nested.py` (`tests` matches testMatch and does get recursed into; `unit` is non-package and doesn't itself contain "test") -> `nosetests -v` **also** collects 0 tests, even though the ancestor `tests/` already matched.
3. Same layout as (2) but with `__init__.py` added to both `tests/` and `unit/` (making them packages) -> `nosetests -v` finds and runs the test -- confirming `ispackage()` alone (not the name) is what rescues it, and that the harness itself isn't just broadly failing to discover things.

This matches nose's own source exactly: both `Selector.wantFile` and `Selector.wantDirectory` do `tail = os.path.basename(...)` then `self.matches(tail)` -- basename only, never the full path. So this repo's `_wants_file`/`_wants_directory` (basename-only) are already faithful to nose 1.3.7's real, verified behavior; switching to full-path matching would make discovery *more permissive* than real nose (recursing into directories nose itself would never enter) and start bin-packing weight budget onto files nose would never actually call `wantFile()` on.

Left this unchanged, and added `NestedNonPackageDirectoryTest` (covering both this exact non-package case and the package-rescues-it case) in the latest commit, to pin down the verified boundary so a future well-intentioned fix here doesn't quietly regress it.

(Separately, the sibling low-severity symlink-duplication finding was real and is fixed in the same commit.)
